src/main/java/com/epam/esm/service/GiftCertificateService.java

```diff
@@ -0,0 +1,133 @@
+    }
+
+    public List<GiftCertificateDto> getGiftCertificates(String tagName, String giftCertificateName, String description, String sortByName, String sortByDate) {
+        List<GiftCertificate> giftCertificates = giftCertificateDao.getGiftCertificates(tagName, giftCertificateName, description, sortByName, sortByDate);
+        List<GiftCertificateDto> giftCertificateDtos = new ArrayList<>();
+
+        giftCertificates.stream().forEach(giftCertificate -> giftCertificateDtos.add(dtoMapper.map(giftCertificate)));
+        return giftCertificateDtos;
+    }
+
+    public GiftCertificateDto getGiftCertificate(long id) throws EntityNotExistsException {
+        Optional<GiftCertificate> optionalGiftCertificate = giftCertificateDao.findById(id);
+        if (optionalGiftCertificate.isEmpty()) {
+            throw new EntityNotExistsException();
+        }
+
+        return dtoMapper.map(optionalGiftCertificate.get());
+    }
+
+    @Transactional
+    public void deleteEntity(long id) throws EntityNotExistsException {
+        if (giftCertificateDao.findById(id).isEmpty()) {
+            throw new EntityNotExistsException();
+        }
+
+        giftCertificateTagDao.deleteGiftCertificateById(id);
+        giftCertificateDao.deleteById(id);
+    }
+
+    @Transactional
+    public GiftCertificateDto updateGiftCertificate(long id, GiftCertificateDto giftCertificateDto) throws EntityNotExistsException, BadEntityException {
+        if (giftCertificateDao.findById(id).isEmpty()) {
+            throw new EntityNotExistsException();
+        }
+
+        GiftCertificate giftCertificate = dtoMapper.unmap(giftCertificateDto);
+
+        if (!validator.validateUpdating(giftCertificate)) {
+            throw new BadEntityException();
+        }
+
+        giftCertificate.setLastUpdateDate(DATE_FORMAT.format(new Date()));
+
+        checkForTags(giftCertificate.getTags());
+        giftCertificateDao.updateGiftCertificate(id, giftCertificate);
+
+        if (!giftCertificate.getTags().isEmpty()) {
+            connectCertificatesAndTags(id, giftCertificate);
+        }
+        return getGiftCertificate(id);
+    }
+
+    @Transactional
+    public GiftCertificateDto createGiftCertificate(GiftCertificateDto giftCertificateDto) throws BadEntityException, EntityNotExistsException {
+        GiftCertificate giftCertificate = dtoMapper.unmap(giftCertificateDto);
+
+        if (!validator.validateCreating(giftCertificate)) {
+            throw new BadEntityException();
+        }
+
+        giftCertificate.setCreationDate(DATE_FORMAT.format(new Date()));
+        giftCertificate.setLastUpdateDate(DATE_FORMAT.format(new Date()));
+
+        checkForTags(giftCertificate.getTags());
+        long lastInsertedId = giftCertificateDao.create(giftCertificate);
+        giftCertificateTagDao.createConnections(lastInsertedId, giftCertificate.getTags());
+
+        return getGiftCertificate(lastInsertedId);
+    }
+
+    private void checkForTags(Set<Tag> newTags) {
```

Comment:
As we've discussed we should return `Set<Tag>` here that is going to contain all the tags that were initalized. Can be changed to smth like this:
```
private Set<Tag> initializeTags(Set<String> incomingTags) {
        return incomingTags.stream()
            .map(tagName -> {
                Optional<Tag> existingTag = tagDao.findTagByName(tagName);
                if (existingTag.isPresent()) {
                    return existingTag.get();
                }
                else {
                    Tag newTag = new Tag(tagName);
                    long tagId = tagDao.create(newTag);
                    newTag.setId(tagId);
                    return newTag;
                }
            })
            .collect(Collectors.toSet());
    }
```

Reply:
The method initializeTags() now returns a Set<Tag>